src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs

```diff
@@ -0,0 +1,628 @@
+
+        private readonly static Type[] s_arrayConstructorParameters = new Type[] { Globals.TypeOfInt };
+        private readonly static object[] s_arrayConstructorArguments = new object[] { 32 };
+
+        public object ReflectionReadClass(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, XmlDictionaryString[] memberNames, XmlDictionaryString[] memberNamespaces, ClassDataContract classContract)
+        {
+            object obj = CreateObject(classContract);
+            context.AddNewObject(obj);
+            InvokeOnDeserializing(context, classContract, obj);
+
+            ReflectionReadMembers(xmlReader, context, memberNames, memberNamespaces, classContract, ref obj);
+            obj = ResolveAdapterObject(obj, classContract);
+
+            InvokeOnDeserialized(context, classContract, obj);
+
+            return obj;
+        }
+
+        public object ReflectionReadCollection(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, XmlDictionaryString collectionItemName, XmlDictionaryString collectionItemNamespace, CollectionDataContract collectionContract)
+        {
+            return ReflectionReadCollectionCore(xmlReader, context, collectionItemName, collectionItemNamespace, collectionContract);
+        }
+
+        public void ReflectionReadGetOnlyCollection(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, XmlDictionaryString collectionItemName, XmlDictionaryString collectionItemNamespace, CollectionDataContract collectionContract)
+        {
+            object resultCollection = context.GetCollectionMember();
+            if (ReflectionReadSpecialCollection(xmlReader, context, collectionContract, resultCollection))
+            {
+                return;
+            }
+
+            if (xmlReader.IsStartElement(collectionItemName, collectionItemNamespace))
+            {
+
+                if (resultCollection == null)
+                {
+                    XmlObjectSerializerReadContext.ThrowNullValueReturnedForGetOnlyCollectionException(collectionContract.UnderlyingType);
+                }
+
+                string itemNameStr = collectionContract.ItemName;
+                string itemNsStr = collectionContract.StableName.Namespace;
+
+                ReflectionReadValueDelegate reflectionReadValueDelegate = null;
+                CollectionSetItemDelegate collectionSetItemDelegate = null;
+                for (int index = 0; index < int.MaxValue; index++)
+                {
+                    if (xmlReader.IsStartElement(collectionItemName, collectionItemNamespace))
+                    {
+                        context.IncrementItemCount(1);
+                        object collectionItem = ReflectionReadCollectionItem(xmlReader, context, collectionContract, collectionContract.ItemType, itemNameStr, itemNsStr, ref reflectionReadValueDelegate);
+                        if (collectionSetItemDelegate == null)
```

Comment:
You already know that xmlReader.IsStartElement will return true in the first loop iteration as the only way to get here is for the if check on line 57 to have returned true. This means you are guaranteed to always hit this check. Move the check and assignment to collectionSetItemDelegate out of the loop and unconditionally set it. This way you avoid an if check on every iteration and make the loop tighter.